servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java

```diff
@@ -74,51 +68,63 @@ private ProxyFilter(final ConnectionFactory<ResolvedAddress, C> delegate) {
-            return delegate().newConnection(resolvedAddress).flatMap(c ->
-                // We currently only have access to a StreamingHttpRequester, which means we are forced to provide an
-                // HttpExecutionStrategy. Because we can't be sure if there is any blocking code in the connection
-                // filters we use the default strategy which should offload everything to be safe.
-                c.request(defaultStrategy(),
-                            reqRespFactory.connect(connectAddress).addHeader(CONTENT_LENGTH, ZERO))
-                 .flatMap(response -> {
-                if (SUCCESSFUL_2XX.contains(response.status())) {
-                    final Channel channel = ((NettyConnectionContext) c.connectionContext()).nettyChannel();
-                    final SingleSource.Processor<C, C> processor = newSingleProcessor();
+            return delegate().newConnection(resolvedAddress).flatMap(c -> {
+                try {
+                    // We currently only have access to a StreamingHttpRequester, which means we are forced to provide
+                    // an HttpExecutionStrategy. Because we can't be sure if there is any blocking code in the
+                    // connection filters we use the default strategy which should offload everything to be safe.
+                    return c.request(defaultStrategy(), c.connect(connectAddress).addHeader(CONTENT_LENGTH, ZERO))
+                            .flatMap(response -> handleConnectResponse(c, response))
+                            // Close recently created connection in case of any error while it connects to the proxy
+                            // or cancellation:
+                            .recoverWith(t -> c.closeAsync().concat(failed(t)))
+                            .whenCancel(() -> c.closeAsync().subscribe());
+                } catch (Throwable t) {
+                    return c.closeAsync().concat(failed(t));
+                }
+            });
+        }
+    }
 
-                    channel.pipeline().addLast(new ChannelInboundHandlerAdapter() {
-                        @Override
-                        public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt) {
-                            if (evt instanceof SslHandshakeCompletionEvent) {
-                                SslHandshakeCompletionEvent event = (SslHandshakeCompletionEvent) evt;
-                                if (event.isSuccess()) {
-                                    processor.onSuccess(c);
-                                } else {
-                                    processor.onError(event.cause());
-                                }
-                            }
-                            ctx.fireUserEventTriggered(evt);
-                        }
-                    });
+    private Single<C> handleConnectResponse(final C connection, final StreamingHttpResponse response) {
+        try {
```

Comment:
Moved this logic to the different method because nested `try` blocks make code indentation awful.